crypto/rsa/rsa_pss.c

```diff
@@ -301,6 +301,47 @@ static const RSA_PSS_PARAMS_30 default_RSASSA_PSS_params = {
     1                            /* default trailerField (0xBC) */
 };
 
+/*
+ * If the hash algorithm is Shake128 OR Shake256 then the algorithm
+ * parameters are restricted to use the following parameters.
+ * These values are not saved to ASN1
+ */
+static const RSA_PSS_PARAMS_30 shake128_RSASSA_PSS_params = {
+    NID_shake128,                /* default hashAlgorithm */
+    {
+        NID_mgf1,                /* default maskGenAlgorithm */
```

Comment:
That does not seem right as per [RFC 8702](https://datatracker.ietf.org/doc/html/rfc8702#section-3.2.1)

> In RSASSA-PSS with SHAKEs, the SHAKEs MUST be used natively as the MGF, instead of the MGF1 algorithm that uses the hash function in multiple iterations, as specified in [Appendix B.2.1](https://www.rfc-editor.org/rfc/rfc8017#appendix-B.2.1) of [[RFC8017](https://datatracker.ietf.org/doc/html/rfc8017)]. In other words, the MGF is defined as the SHAKE128 or SHAKE256 with input being the mgfSeed for id-RSASSA-PSS-SHAKE128 and id-RSASSA-PSS-SHAKE256, respectively.

Thus I am expected no NID_mgf1 here.

Reply:
@xnox

Could you give me your opinion on the test vectors located here
https://github.com/usnistgov/ACVP-Server/blob/master/gen-val/json-files/RSA-SigVer-FIPS186-5/internalProjection.json

Look at the SHAKE entries in particular... it seems to allow alg/maskgenalgorithm combinations such as shake-128 with shake-256 and  sha3 with shakeX

The "FIPS 186-5 5.4.1 Mask Generation Functions" does not mention the CMS related changes.. So there is a mismatch between the specs here..

Reply:
Half of the test vectors look fine to me (as they use shake as the mgf function), I am confused about those that imply using MGF1.

Reply:
The test vectors that use Shake + "MGF1" seem to use shake inside the MGF1 function. (They also seem to use 1/2 the security strength currently inside the MGF1 for shake - I dont think this should be supported inside OpenSSL).

Reply:
RFC 8933 unambiguously states that when SHAKE is used in digital signature hash, SHAKE must be used as the MGF function, and not as mgf1(shake). Thus whatever ACVP server states appears to be in violation of RFC 8933.

Also reading FIPS 186 it doesn't explicitly call out SHAKE as allowed in MGF1 (section 5.4.1), in the same manner as it calls out that SHAKE is allowed for message hash (section 5.4(b)).

Imho SHAKE in MGF1 must not be supported by OpenSSL as it would violate RFC 8933. I will separately try to contact NIST about this, for them to clarify this.